File: ios/SCMemorial/CCAvenuePaymentModule.swift
//
//  CCAvenuePaymentModule.swift
//
//  CCAvenue payment — NATIVE INTEGRATION BOUNDARY (iOS).
//
//  ============================================================================
//  STATUS: PLACEHOLDER — NOT PART OF THE XCODE TARGET YET.
//  ============================================================================
//
//  The React Native payment architecture currently runs in MOCK mode
//  (src/config/payment.ts, PAYMENT_MODE = "MOCK"). The mock adapter is fully
//  self-contained in JavaScript and never calls this module.
//
//  This file is the Swift seam where the OFFICIAL CCAvenue iOS SDK will be
//  connected later. It is intentionally NOT added to the Xcode project
//  (SCMemorial.xcodeproj) right now, so it does not affect the build.
//
//  When the official CCAvenue iOS SDK is available and PAYMENT_MODE switches
//  to "CCAVENUE":
//
//    1. Add the official CCAvenue iOS SDK (CocoaPods / SPM) and add this file
//       to the SCMemorial target in SCMemorial.xcodeproj.
//    2. Implement `startPayment` to launch the official CCAvenue payment
//       screen using the merchant configuration (merchantId / accessCode /
//       workingKey / encryptionKey) passed from JavaScript.
//    3. Implement `verifyPayment` if CCAvenue exposes client-side checks
//       (server-side verification should live in the backend).
//
//  The JS side (src/services/payment/RealCCAvenueProvider.ts) already has the
//  matching calls, marked with TODOs.
//
//  NOTE: No official CCAvenue SDK APIs are referenced here because the SDK is
//  not installed in this project yet. Nothing here is an official CCAvenue
//  API or class.

import Foundation
import React

@objc(CCAvenuePayment)
class CCAvenuePaymentModule: NSObject {

  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }

  /// Starts a CCAvenue payment session.
  /// - Parameters:
  ///   - options: JSON dictionary with orderId/amount/currency/billing details.
  ///   - resolver: Resolves with the normalized payment result.
  ///   - rejecter: Rejects when the payment could not be started.
  @objc(startPayment:resolver:rejecter:)
  func startPayment(_ options: NSDictionary,
                    resolver: @escaping RCTPromiseResolveBlock,
                    rejecter: @escaping RCTPromiseRejectBlock) {
    // TODO(real CCAvenue):
    //   let orderId = options["orderId"] as? String ?? ""
    //   let amount = options["amount"] as? String ?? "0"
    //   let currency = options["currency"] as? String ?? "INR"
    //   ... initialize the official CCAvenue SDK with merchant config ...
    //   ... present the CCAvenue payment UI and await the callback ...
    //   ... map the gateway callback into the normalized result dictionary
    //       and call resolver(result) or rejecter(code, message, nil) ...
    rejecter(
      "NOT_CONFIGURED",
      "Real CCAvenue SDK is not configured yet. The app currently runs in MOCK payment mode.",
      nil
    )
  }

  /// Cancels an in-flight CCAvenue payment session if the SDK supports it.
  @objc(cancelPayment:rejecter:)
  func cancelPayment(_ resolver: @escaping RCTPromiseResolveBlock,
                     rejecter: @escaping RCTPromiseRejectBlock) {
    // TODO(real CCAvenue): dismiss/abort the SDK session if possible.
    rejecter(
      "NOT_CONFIGURED",
      "Real CCAvenue SDK is not configured yet. The app currently runs in MOCK payment mode.",
      nil
    )
  }

  /// Placeholder for any client-side transaction status lookups.
  @objc(verifyPayment:transactionId:resolver:rejecter:)
  func verifyPayment(_ orderId: String,
                     transactionId: String,
                     resolver: @escaping RCTPromiseResolveBlock,
                     rejecter: @escaping RCTPromiseRejectBlock) {
    // TODO(real CCAvenue): optional client-side status check.
    rejecter(
      "NOT_CONFIGURED",
      "Real CCAvenue SDK is not configured yet. The app currently runs in MOCK payment mode.",
      nil
    )
  }
}

File: android/app/src/main/java/com/scmemorial/payment/CCAvenuePaymentModule.kt
package com.scmemorial.payment

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap

/**
 * CCAvenuePaymentModule — NATIVE INTEGRATION BOUNDARY.
 *
 * ============================================================================
 * STATUS: PLACEHOLDER — NOT WIRED INTO THE ACTIVE PAYMENT FLOW.
 * ============================================================================
 *
 * The React Native payment architecture currently runs in MOCK mode
 * (src/config/payment.ts, PAYMENT_MODE = "MOCK"). The mock adapter is fully
 * self-contained in JavaScript and does NOT call this module. This file is
 * the Kotlin seam where the OFFICIAL CCAvenue Android SDK will be connected
 * later.
 *
 * When the official CCAvenue SDK is available and PAYMENT_MODE switches to
 * "CCAVENUE":
 *
 *   1. Add the official CCAvenue SDK dependency to android/app/build.gradle.
 *   2. Implement `startPayment` here to launch the CCAvenue payment activity
 *      using the merchant configuration (merchantId / accessCode / workingKey /
 *      encryptionKey) that will be passed from JavaScript.
 *   3. Implement `verifyPayment` if CCAvenue exposes client-side status checks.
 *      (Server-side verification should live in the backend, not the app.)
 *   4. Register this package in MainApplication.kt inside the `packages` list:
 *
 *          add(CCAvenuePaymentPackage())
 *
 * The JS side (src/services/payment/RealCCAvenueProvider.ts) already has the
 * matching calls, marked with TODOs.
 *
 * NOTE: This module intentionally contains NO real CCAvenue SDK calls today.
 * No official CCAvenue SDK classes are referenced because the SDK is not
 * installed in this project yet.
 */
class CCAvenuePaymentModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = NAME

    companion object {
        const val NAME = "CCAvenuePayment"
    }

    /**
     * Starts a CCAvenue payment session.
     *
     * @param options Map with orderId, amount, currency, billingName,
     *                billingEmail, billingPhone (payload from JS).
     * @param promise Resolves with the normalized payment result
     *                (status/orderId/transactionId/amount/currency/message).
     */
    @ReactMethod
    fun startPayment(options: ReadableMap, promise: Promise) {
        // TODO(real CCAvenue):
        //   val orderId = options.getString("orderId") ?: ""
        //   val amount = options.getString("amount") ?: "0"
        //   val currency = options.getString("currency") ?: "INR"
        //   ... initialize the official CCAvenue SDK with the merchant config ...
        //   ... launch the payment UI and wait for the callback ...
        //   ... map the gateway callback into the normalized result map and
        //       call promise.resolve(resultMap) or promise.reject(...) ...
        promise.reject(
            "NOT_CONFIGURED",
            "Real CCAvenue SDK is not configured yet. The app currently runs in MOCK payment mode."
        )
    }

    /**
     * Cancels an in-flight CCAvenue payment session if the SDK supports it.
     */
    @ReactMethod
    fun cancelPayment(promise: Promise) {
        // TODO(real CCAvenue): close/abort the SDK session if possible.
        promise.reject(
            "NOT_CONFIGURED",
            "Real CCAvenue SDK is not configured yet. The app currently runs in MOCK payment mode."
        )
    }

    /**
     * Placeholder for any client-side transaction status lookups.
     * Prefer server-side verification via the backend.
     */
    @ReactMethod
    fun verifyPayment(orderId: String, transactionId: String, promise: Promise) {
        // TODO(real CCAvenue): optional client-side status check.
        promise.reject(
            "NOT_CONFIGURED",
            "Real CCAvenue SDK is not configured yet. The app currently runs in MOCK payment mode."
        )
    }
}
